### PR TITLE
BUG: DatetimeIndex.insert on empty can preserve freq

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_insert.py [hunk 1]
```diff
@@ -24,6 +24,19 @@ def test_insert_invalid_na(self, tz):
         with pytest.raises(TypeError, match="incompatible label"):
             idx.insert(0, np.timedelta64("NaT"))
 
+    def test_insert_empty_preserves_freq(self, tz_naive_fixture):
+        tz = tz_naive_fixture
+        dti = DatetimeIndex([], tz=tz, freq="D")
+        item = Timestamp("2017-04-05").tz_localize(tz)
+
+        result = dti.insert(0, item)
+        assert result.freq == dti.freq
```
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/test_insert.py:33]: IIRC you mentioned that `assert_index_equal` doesn't compare freq i.e. we can't just simply use `assert_index_equal` here?

Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_insert.py:33]: correct.  ive got a branch that adds a freq check, am whittling down the actual bugs it identifies before making a PR for it.

File: pandas/tests/indexes/timedeltas/test_insert.py [hunk 2]
```diff
@@ -96,6 +96,11 @@ def test_insert_empty(self):
         idx = timedelta_range("1 Day", periods=3)
         td = idx[0]
 
-        idx[:0].insert(0, td)
-        idx[:0].insert(1, td)
-        idx[:0].insert(-1, td)
+        result = idx[:0].insert(0, td)
```
Comment (mroeschke) [inline pandas/tests/indexes/timedeltas/test_insert.py:99]: Worth parameterizing?

Comment (jbrockmendel) [reply pandas/tests/indexes/timedeltas/test_insert.py:99]: i dont think so.  more interested in trying to parametrize across DTI/TDI, but thats out of scope

**Review by mroeschke — APPROVED** [review]

File: pandas/tests/indexes/datetimes/test_insert.py [hunk 3]
```diff
@@ -24,6 +24,19 @@ def test_insert_invalid_na(self, tz):
         with pytest.raises(TypeError, match="incompatible label"):
             idx.insert(0, np.timedelta64("NaT"))
 
+    def test_insert_empty_preserves_freq(self, tz_naive_fixture):
+        tz = tz_naive_fixture
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_insert.py:28]: can you add this PR number here

File: pandas/tests/indexes/timedeltas/test_insert.py (hunk 2 above)
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_insert.py:99]: can you add the PR number

**Review by jreback — COMMENTED** [review]
lgtm. can you add a whatsnew, merge on green.

**Pull request merged**